src/Reflection/ClassNameHelper.php

```diff
@@ -3,15 +3,23 @@
 namespace PHPStan\Reflection;
 
 use Nette\Utils\Strings;
+use function array_key_exists;
 use function ltrim;
 
 final class ClassNameHelper
 {
 
+	/** @var array<string, bool> */
+	private static array $checked = [];
+
 	public static function isValidClassName(string $name): bool
 	{
-		// from https://stackoverflow.com/questions/3195614/validate-class-method-names-with-regex#comment104531582_12011255
-		return Strings::match(ltrim($name, '\\'), '/^[a-zA-Z_\x80-\xff][a-zA-Z0-9_\x80-\xff]*(\\\\[a-zA-Z_\x80-\xff][a-zA-Z0-9_\x80-\xff]*)*$/') !== null;
+		if (!array_key_exists($name, self::$checked)) {
+			// from https://stackoverflow.com/questions/3195614/validate-class-method-names-with-regex#comment104531582_12011255
+			self::$checked[$name] = Strings::match(ltrim($name, '\\'), '/^[a-zA-Z_\x80-\xff][a-zA-Z0-9_\x80-\xff]*(\\\\[a-zA-Z_\x80-\xff][a-zA-Z0-9_\x80-\xff]*)*$/') !== null;
```

Comment:
How much faster https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/v3.75.0/src/DocBlock/TypeExpression.php#L177-L179 this regex is?

Regexes are usually fast.

`(...*)*` in theory can be very slow compared to `(...*+)*`.

Also the trim can be easily integrated into the regex.